lib/pbench/server/filetree.py

```diff
@@ -298,20 +326,55 @@ def unpack(self, incoming: Path, results: Path):
         ensured by calling this indirectly through the Controller class unpack
         method.
 
-        TODO: This is a prototype for testing, and doesn't actually unpack
-        the tarball as we're going to be relying on the current unpack pipeline
-        command for some time.
-
         Args:
             incoming: Controller's directory in the INCOMING tree
             results: Controller's directory in the RESULTS tree
+
+        Raises:
+            TarballUnpackError if an exception/failure occurs while unpacking
+            the tarball.
         """
-        self.logger.warning("Tarball unpack is not yet implemented")
-        unpacked = incoming / self.name
-        unpacked.mkdir(parents=True)  # Just create an empty directory for now
-        self.unpacked = unpacked
+        unpacked = incoming / f"{self.name}.unpack"
+        unpacked.mkdir(parents=True)
+
+        try:
+            process = subprocess.run(
+                ["tar", "-xf", self.tarball_path],
+                cwd=unpacked,
+                stdin=subprocess.DEVNULL,
+                capture_output=True,
+                text=True,
+            )
+        except Exception as exc:
+            raise TarballUnpackError(self.tarball_path.name, str(exc)) from exc
+        else:
+            if process.returncode != 0:
+                raise TarballUnpackError(
+                    self.tarball_path.name,
+                    f"tar exited with status {process.returncode}:  {process.stderr.strip()!r}",
+                )
+
+        command = "find unpacked/ \( -type d -exec chmod ugo+rx '{}' \; \) -o \( -type f -exec chmod ugo+r '{}' \; \)"
+        try:
+            process = subprocess.run(
+                command,
+                shell=True,
+                stdin=subprocess.DEVNULL,
+                capture_output=True,
+                text=True,
+            )
```

Comment:
There's a small inconsistency between the invocation of `tar` and the invocation of `find` which, if you reconcile it, might make unit testing more straightforward:  for the `tar` command, you're specifying the current working directory and having `tar` operate relative to it; for the `find` command, you're specifying the target directory explicitly without specifying the current working directory (although, awkwardly, I think you're specifying a relative path for the target, which means that we're depending on whatever `subprocess.run()` picks as the default working directory).

I recommend being explicit and consistent.  The simplest approach would be to explicitly pass the `cwd` parameter to `subprocess.run()` the same way in both invocations, and to have both `tar` and `find` target the current working directory (i.e., the `tar` command is doing this now; the `find` command would be changed to target `.`).

If you pass the `cwd` parameter in both cases, then in the unit test the mock can check it unconditionally, without having to consider which command is being run, and it won't have to grub around in the argument list looking for the target.

Also, I recommend omitting the `shell=true` argument.  This puts everything under your direct control so you shouldn't need all the escapes:
```
            command = (
                "find . ( -type d -exec chmod ugo+rx '{}' + ) "
                "-o ( -type f -exec chmod ugo+r '{}' + )"
            )
            process = subprocess.run(
                command.split(),
                cwd=unpacked,
                stdin=subprocess.DEVNULL,
                capture_output=True,
                text=True,
            )
```

Reply:
I've noticed the `tar -xf` several times and passed over it because it's somewhat "canonical", even though it's very different from what the current bash script does. I've been fiddling with the current bash script in the context of trying to get my new pipeline working, even though I expect that to be obsolete by the time I merge... the experience does, however, give me a bit of perspective on this I've been neglecting to share.

The bash does

```bash
tar --extract --no-same-owner --touch --delay-directory-restore --file="${result}" --force-local --directory="${incoming}.unpack
```

* `--extract` is just the `-x`
* `--no-same-owner` is essential: the tarball will normally be created by `root`, but we need the unpacked files to be owned by `pbench` (the user as which our cron jobs run). We could use `--owner=pbench --group=pbench` instead, but `--no-same-owner` doesn't encode an assumed user/group.
* `--touch`, interestingly, omits the last-modified time; I infer from the name that the last-modified time will default to the local creation time of the file. (As if you'd followed with a `touch` command.) That's an interesting choice, and I don't really see why we need it. (Even the current `cull` command constructs a timestamp based on the date fields in the tarball *name*, without reference to actual file timestamps... though this always seemed a bit weird to me.)
* `--delay-directory-restore` defers setting mode and modification times on directories until *after* extracting everything, and is recommended for tarballs with "unusual member ordering", and I can't see why it really matters to us.
* `--file` is the `-f`
* `--force-local` ... apparently `tar` will actually attempt remote file access if the name of the tarball has a `:` ... this suppresses that behavior; e.g., for our `EC2::` prefixes, so I think we need this. Maybe; depending on what we do with the EC2 satellite. (I'd rather not preserve those prefixes, whatever we do, but for now we probably need this.)
* `--directory` causes an implicit `chmod` before the extract; so one or the other is redundant, but I think I prefer `--directory` (`-c` for short) since it puts all the context in one place.

__In summary__

I think we need at least `tar --no-same-owner --force-local` in addition to the `-xf` or `--file --extract`. I'd like to have `--directory` instead of the separate `chmod` for clarity. The others are debatable: maybe @ndokos and/or @portante can weigh in on the historical background for `--touch` and `--delay-directory-restore`.

Reply:
> `--directory` causes an implicit `chmod` before the extract; so one or the other is redundant, but I think I prefer `--directory` (`-c` for short) since it puts all the context in one place.

@dbutenhof, I'm confused by this.  I think `--directory` (whose short form is `-C`, not `-c`, which does something else) causes an implicit _`cd`_, not `chmod`.

Beyond that...

I can see how `--delay-directory-restore` could be important, since we don't really know anything about the tool used to create the tarball:  we wouldn't want `tar` to create a directory which it could not then write files into.

I could see how `--touch` might be worthwhile, if it does indeed work as you suppose.

`--no-same-owner` appears to be the default, assuming that the tool is not running as a superuser; however, being explicit is a good idea.

Similarly, `--force-local` is just a good idea (since we expect all the files to be local).

Reply:
> @dbutenhof, I'm confused by this. I think `--directory` (whose short form is `-C`, not `-c`, which does something else) causes an implicit _`cd`_, not `chmod`.

Ha; typos aren't confusing, they're just misleading. 😆 

But now I'm going to make this really confusing by editing the original typo so it's no longer confusing on its own...

Reply:
> But now I'm going to make this really confusing by editing the original typo so it's no longer confusing on its own...

🎉 

Now, back to the discussion at hand....

> I think I prefer `--directory` (`-C` for short) since it puts all the context in one place.

I initially held this preference as well, until I started thinking about what @riya-17 was going to have to do to unit-test this stuff.  Then I decided that it would be better to provide `cwd=uploaded` on another line of the same invocation which would otherwise specify `-C` and `uploaded` on two previous lines.  Either way, the context is all in "one place" (it's just in a different parameter of the same function call rather than in additional elements in the list in the first one), but providing it as a separate parameter makes it easy to locate for the assertion in the mock (if it is instead part of the command args list, then we either have to somehow "know" where it is, or we have to do some sort of search for it, neither of which is particularly satisfying).